File: ref/bitunix-demo/Demo/Golang/open_api_http_future_private.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
)

// OpenApiHttpFuturePrivate future private api client
type OpenApiHttpFuturePrivate struct {
	apiKey    string
	secretKey string
	baseURL   string
	client    *http.Client
}

// NewOpenApiHttpFuturePrivate create new future private api client
func NewOpenApiHttpFuturePrivate(config *Config) *OpenApiHttpFuturePrivate {
	return &OpenApiHttpFuturePrivate{
		apiKey:    config.Credentials.APIKey,
		secretKey: config.Credentials.SecretKey,
		baseURL:   config.HTTP.URIPrefix,
		client:    &http.Client{},
	}
}

// handleResponse handle future private api response
func (c *OpenApiHttpFuturePrivate) handleResponse(resp *http.Response) (map[string]interface{}, error) {
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("HTTP Error: %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var result map[string]interface{}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, err
	}

	if code, ok := result["code"].(float64); ok && code != 0 {
		msg, _ := result["msg"].(string)
		return nil, fmt.Errorf("API Error: %v - %s", code, msg)
	}

	if data, ok := result["data"].(map[string]interface{}); ok {
		return data, nil
	}

	return nil, fmt.Errorf("invalid response format")
}

// BuildQueryString build get request query string
func BuildQueryStringPrivate(params map[string]string) string {
	if len(params) == 0 {
		return ""
	}

	var queryString strings.Builder
	first := true
	for k, v := range params {
		if !first {
			queryString.WriteString("&")
		}
		queryString.WriteString(k)
		queryString.WriteString("=")
		queryString.WriteString(v)
		first = false
	}
	return "?" + queryString.String()
}

// GetAccount get account info
func (c *OpenApiHttpFuturePrivate) GetAccount(marginCoin string) (map[string]interface{}, error) {
	if marginCoin == "" {
		marginCoin = "USDT"
	}
	params := make(map[string]string)
	params["marginCoin"] = marginCoin

	headers := getAuthHeadersPrivate(c.apiKey, c.secretKey, params, "")

	// build full url
	queryStr := BuildQueryStringPrivate(params)
	req, err := http.NewRequest("GET", c.baseURL+"/api/v1/futures/account"+queryStr, nil)
	if err != nil {
		return nil, err
	}

	for k, v := range headers {
		req.Header.Set(k, v)
	}

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	return c.handleResponse(resp)
}

// PlaceOrder place order
func (c *OpenApiHttpFuturePrivate) PlaceOrder(params map[string]interface{}) (map[string]interface{}, error) {
	jsonData, err := json.Marshal(params)
	if err != nil {
		return nil, err
	}

	headers := getAuthHeadersPrivate(c.apiKey, c.secretKey, map[string]string{}, string(jsonData))

	req, err := http.NewRequest("POST", c.baseURL+"/api/v1/futures/trade/place_order", bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, err
	}

	for k, v := range headers {
		req.Header.Set(k, v)
	}

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	return c.handleResponse(resp)
}

// CancelOrders cancel orders
func (c *OpenApiHttpFuturePrivate) CancelOrders(symbol string, orderList []map[string]string) (map[string]interface{}, error) {
	data := map[string]interface{}{
		"symbol":    symbol,
		"orderList": orderList,
	}

	jsonData, err := json.Marshal(data)
	if err != nil {
		return nil, err
	}

	headers := getAuthHeadersPrivate(c.apiKey, c.secretKey, map[string]string{}, string(jsonData))

	req, err := http.NewRequest("POST", c.baseURL+"/api/v1/futures/trade/cancel_orders", bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, err
	}

	for k, v := range headers {
		req.Header.Set(k, v)
	}

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	return c.handleResponse(resp)
}

// GetHistoryOrders get history orders
func (c *OpenApiHttpFuturePrivate) GetHistoryOrders(symbol string) (map[string]interface{}, error) {
	params := make(map[string]string)
	if symbol != "" {
		params["symbol"] = symbol
	}

	headers := getAuthHeadersPrivate(c.apiKey, c.secretKey, params, "")

	// build url query params
	queryStr := BuildQueryStringPrivate(params)
	req, err := http.NewRequest("GET", c.baseURL+"/api/v1/futures/trade/get_history_orders"+queryStr, nil)
	if err != nil {
		return nil, err
	}

	for k, v := range headers {
		req.Header.Set(k, v)
	}

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	return c.handleResponse(resp)
}

// GetHistoryPositions get history positions
func (c *OpenApiHttpFuturePrivate) GetHistoryPositions(symbol string) (map[string]interface{}, error) {
	params := make(map[string]string)
	if symbol != "" {
		params["symbol"] = symbol
	}

	headers := getAuthHeadersPrivate(c.apiKey, c.secretKey, params, "")

	// build url query params
	queryStr := BuildQueryStringPrivate(params)
	req, err := http.NewRequest("GET", c.baseURL+"/api/v1/futures/position/get_history_positions"+queryStr, nil)
	if err != nil {
		return nil, err
	}

	for k, v := range headers {
		req.Header.Set(k, v)
	}

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	return c.handleResponse(resp)
}

// getAuthHeadersPrivate get auth headers
func getAuthHeadersPrivate(apiKey string, secretKey string, queryParam map[string]string, body string) map[string]string {
	queryString := SortParams(queryParam)
	headers := GetAuthHeaders(apiKey, secretKey, queryString, body)
	headers["language"] = "en-US"
	headers["Content-Type"] = "application/json"
	return headers
}

// HttpPrivateExampleUsage main function
func HttpPrivateExampleUsage() {
	// load config
	config, err := LoadConfig()
	if err != nil {
		fmt.Printf("load config failed: %v\n", err)
		return
	}

	// create client
	client := NewOpenApiHttpFuturePrivate(config)

	// get account info
	account, err := client.GetAccount("USDT")
	if err != nil {
		fmt.Printf("get account info failed: %v\n", err)
		return
	}
	fmt.Printf("account info: %+v\n", account)

	// get history positions
	positions, err := client.GetHistoryPositions("BTCUSDT")
	if err != nil {
		fmt.Printf("get history positions failed: %v\n", err)
		return
	}
	fmt.Printf("history positions: %+v\n", positions)

	// get history orders
	orders, err := client.GetHistoryOrders("BTCUSDT")
	if err != nil {
		fmt.Printf("get history orders failed: %v\n", err)
		return
	}
	fmt.Printf("history orders: %+v\n", orders)

	/*
		WARNING!!! This is example code for placing and canceling orders. If you are using a real account,
		please be cautious when uncommenting for testing, as any financial losses will be your responsibility.
	*/
	// // Example order placement (limit order)
	// order, err := client.PlaceOrder(map[string]interface{}{
	// 	"symbol":       "BTCUSDT",
	// 	"side":         "BUY",
	// 	"orderType":    "LIMIT",
	// 	"qty":          "0.5",
	// 	"price":        "60000",
	// 	"tradeSide":    "OPEN",
	// 	"effect":       "GTC",
	// 	"reduceOnly":   false,
	// 	"clientId":     time.Now().Format("20060102150405"),
	// 	"tpPrice":      "61000",
	// 	"tpStopType":   "MARK",
	// 	"tpOrderType":  "LIMIT",
	// 	"tpOrderPrice": "61000.1",
	// })
	// if err != nil {
	// 	fmt.Printf("place order failed: %v\n", err)
	// 	return
	// }
	// fmt.Printf("place order result: %+v\n", order)

	// // Example order cancellation
	// if order != nil && order["orderId"] != nil {
	// 	cancel_result, err := client.CancelOrders("BTCUSDT", []map[string]string{
	// 		{"orderId": order["orderId"].(string)},
	// 		{"clientId": order["clientId"].(string)},
	// 	})
	// 	if err != nil {
	// 		fmt.Printf("cancel order failed: %v\n", err)
	// 		return
	// 	}
	// 	fmt.Printf("cancel order result: %+v\n", cancel_result)
	// }

}

File: ref/bitunix-demo/Demo/PHP/open_api_http_future_private.php
<?php

require_once 'config.php';
require_once 'error_codes.php';
require_once 'open_api_http_sign.php';

/**
 * OpenAPI http future private api client
 */
class OpenApiHttpFuturePrivate {
    private Config $config;
    private string $apiKey;
    private string $secretKey;
    private string $baseUrl;
    private array $defaultHeaders;

    /**
     * initialize OpenApiHttpFuturePrivate class
     * 
     * @param Config $config config object
     */
    public function __construct(Config $config) {
        $this->config = $config;
        $this->apiKey = $config->getApiKey();
        $this->secretKey = $config->getSecretKey();
        $this->baseUrl = $config->getUriPrefix();
        
        // set common request headers
        $this->defaultHeaders = [
            "language" => "en-US",
            "Content-Type" => "application/json"
        ];
    }

    /**
     * handle response
     * 
     * @param array $response response data
     * @return array processed data
     * @throws Exception when the response status code is not 200 or the business status code is not 0
     */
    private function handleResponse(array $response): array {
        if ($response['http_code'] !== 200) {
            throw new Exception("HTTP Error: " . $response['http_code']);
        }

        $data = json_decode($response['body'], true);
        if ($data['code'] !== 0) {
            $error = ErrorCode::getByCode($data['code']);
            if ($error) {
                throw new Exception(json_encode($error));
            }
            throw new Exception("Unknown Error: {$data['code']} - {$data['msg']}");
        }

        return $data['data'];
    }

    /**
    * send http request
     * 
     * @param string $method request method
     * @param string $url request url
     * @param array $params query params
     * @param array $data request body data
     * @return array response data
     */
    private function sendRequest(string $method, string $url, array $params = [], array $data = []): array {
        $ch = curl_init();
        
        // build full url 
        if (!empty($params)) {
            $url .= '?' . http_build_query($params);
        }
        
        // set curl options
        curl_setopt($ch, CURLOPT_URL, $url);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        
        // prepare request headers
        $headers = $this->defaultHeaders;
        
        if ($method === 'POST') {
            curl_setopt($ch, CURLOPT_POST, true);
            if (!empty($data)) {
                $jsonData = json_encode($data);
                curl_setopt($ch, CURLOPT_POSTFIELDS, $jsonData);
                $authHeaders = OpenApiHttpSign::getAuthHeaders($this->apiKey, $this->secretKey, [], $jsonData);
            } else {
                $authHeaders = OpenApiHttpSign::getAuthHeaders($this->apiKey, $this->secretKey);
            }
        } else {
            $authHeaders = OpenApiHttpSign::getAuthHeaders($this->apiKey, $this->secretKey, $params);
        }
        
        // merge all headers
        $headers = array_merge($headers, $authHeaders);
        
        // set request headers
        $headerArray = [];
        foreach ($headers as $key => $value) {
            $headerArray[] = "$key: $value";
        }
        curl_setopt($ch, CURLOPT_HTTPHEADER, $headerArray);
        
        // execute request
        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);
        
        return [
            'http_code' => $httpCode,
            'body' => $response
        ];
    }

    /**
     * get account info
     * 
     * @param string $marginCoin margin coin, default USDT
     * @return array account info
     */
    public function getAccount(string $marginCoin = "USDT"): array {
        $url = $this->baseUrl . "/api/v1/futures/account";
        $params = [
            "marginCoin" => $marginCoin
        ];
        
        $response = $this->sendRequest('GET', $url, $params);
        return $this->handleResponse($response);
    }

    /**
     * place order
     * 
     * @param string $symbol symbol
     * @param string $side side, BUY or SELL
     * @param string $orderType order type, LIMIT or MARKET
     * @param string $qty quantity
     * @param string|null $price price, required when orderType is LIMIT
     * @param string|null $positionId position id
     * @param string $tradeSide trade side, OPEN (open position) or CLOSE (close position), default OPEN
     * @param string $effect order effective time, default GTC
     * @param bool $reduceOnly reduce only
     * @param string|null $clientId client order id
     * @param string|null $tpPrice take profit price
     * @param string|null $tpStopType take profit stop type, MARK (mark price) or LAST (last price)
     * @param string|null $tpOrderType take profit order type, LIMIT or MARKET
     * @param string|null $tpOrderPrice take profit order price
     * @return array order info
     */
    public function placeOrder(
        string $symbol,
        string $side,
        string $orderType,
        string $qty,
        ?string $price = null,
        ?string $positionId = null,
        string $tradeSide = "OPEN",
        string $effect = "GTC",
        bool $reduceOnly = false,
        ?string $clientId = null,
        ?string $tpPrice = null,
        ?string $tpStopType = null,
        ?string $tpOrderType = null,
        ?string $tpOrderPrice = null
    ): array {
        $url = $this->baseUrl . "/api/v1/futures/trade/place_order";
        
        $data = [
            "symbol" => $symbol,
            "side" => $side,
            "orderType" => $orderType,
            "qty" => $qty,
            "tradeSide" => $tradeSide,
            "effect" => $effect,
            "reduceOnly" => $reduceOnly
        ];
        
        if ($price !== null) {
            $data["price"] = $price;
        }
        if ($positionId !== null) {
            $data["positionId"] = $positionId;
        }
        if ($clientId !== null) {
            $data["clientId"] = $clientId;
        }
        if ($tpPrice !== null) {
            $data["tpPrice"] = $tpPrice;
        }
        if ($tpStopType !== null) {
            $data["tpStopType"] = $tpStopType;
        }
        if ($tpOrderType !== null) {
            $data["tpOrderType"] = $tpOrderType;
        }
        if ($tpOrderPrice !== null) {
            $data["tpOrderPrice"] = $tpOrderPrice;
        }
        
        $response = $this->sendRequest('POST', $url, [], $data);
        return $this->handleResponse($response);
    }

    /**
     * cancel orders
     * 
     * @param string $symbol symbol
     * @param array $orderList order list, each order contains orderId or clientId
     * @return array cancel order result
     */
    public function cancelOrders(string $symbol, array $orderList): array {
        $url = $this->baseUrl . "/api/v1/futures/trade/cancel_orders";
        
        $data = [
            "symbol" => $symbol,
            "orderList" => $orderList
        ];
        
        $response = $this->sendRequest('POST', $url, [], $data);
        return $this->handleResponse($response);
    }

    /**
     * get history orders
     * 
     * @param string|null $symbol symbol, null for all symbols
     * @return array history orders list
     */
    public function getHistoryOrders(?string $symbol = null): array {
        $url = $this->baseUrl . "/api/v1/futures/trade/get_history_orders";
        $params = [];
        if ($symbol) {
            $params["symbol"] = $symbol;
        }
        
        $response = $this->sendRequest('GET', $url, $params);
        return $this->handleResponse($response);
    }

    /**
     * get history positions
     * 
     * @param string|null $symbol symbol, null for all symbols
     * @return array history positions
     */
    public function getHistoryPositions(?string $symbol = null): array {
        $url = $this->baseUrl . "/api/v1/futures/position/get_history_positions";
        $params = [];
        if ($symbol) {
            $params["symbol"] = $symbol;
        }
        
        $response = $this->sendRequest('GET', $url, $params);
        return $this->handleResponse($response);
    }

    /**
     * get current positions
     * 
     * @param string|null $symbol symbol, null for all symbols
     * @return array current positions
     */
    public function getCurrentPositions(?string $symbol = null): array {
        $url = "/api/v1/futures/position/get_pending_positions";
        $params = [];
        if ($symbol) {
            $params["symbol"] = $symbol;
        }
        
        $response = $this->sendRequest('GET', $url, $params);
        return $this->handleResponse($response);
    }
}

// example usage
if (php_sapi_name() === 'cli') {
    try {
        // load config
        $config = new Config();
        
        // create client
        $client = new OpenApiHttpFuturePrivate($config);
        
        // get account info
        $account = $client->getAccount();
        echo "Account info: " . json_encode($account, JSON_PRETTY_PRINT) . "\n";
        
        // get current positions
        $currentPositions = $client->getCurrentPositions("BTCUSDT");
        echo "Current positions: " . json_encode($currentPositions, JSON_PRETTY_PRINT) . "\n";
        
        // get history positions
        $historyPositions = $client->getHistoryPositions("BTCUSDT");
        echo "History positions: " . json_encode($historyPositions, JSON_PRETTY_PRINT) . "\n";
        
        // get history orders
        $historyOrders = $client->getHistoryOrders("BTCUSDT");
        echo "History orders: " . json_encode($historyOrders, JSON_PRETTY_PRINT) . "\n";
        
        /*
        WARNING!!! This is example code for placing and canceling orders. If you are using a real account,
        please be cautious when uncommenting for testing, as any financial losses will be your responsibility.
        */
        /*
        // place order example (limit order)
        $order = $client->placeOrder(
            symbol: "BTCUSDT",
            side: "BUY",
            orderType: "LIMIT",
            qty: "0.5",
            price: "60000",
            tradeSide: "OPEN",
            effect: "GTC",
            reduceOnly: false,
            clientId: date("YmdHis"),
            tpPrice: "61000",
            tpStopType: "MARK",
            tpOrderType: "LIMIT",
            tpOrderPrice: "61000.1"
        );
        echo "Place order result: " . json_encode($order, JSON_PRETTY_PRINT) . "\n";
        
        // cancel order example
        if ($order && isset($order["orderId"])) {
            $cancelResult = $client->cancelOrders("BTCUSDT", [
                ["orderId" => $order["orderId"]]
            ]);
            echo "Cancel order result: " . json_encode($cancelResult, JSON_PRETTY_PRINT) . "\n";
        }
        */
        
    } catch (Exception $e) {
        echo "Error: " . $e->getMessage() . "\n";
    }
} 